src/GovUk.Education.ExploreEducationStatistics.Admin/Services/Public.Data/DataSetService.cs

```diff
@@ -0,0 +1,119 @@
+using GovUk.Education.ExploreEducationStatistics.Common.Extensions;
+using GovUk.Education.ExploreEducationStatistics.Common.Model;
+using GovUk.Education.ExploreEducationStatistics.Common.Services.Interfaces.Security;
+using GovUk.Education.ExploreEducationStatistics.Common.ViewModels;
+using GovUk.Education.ExploreEducationStatistics.Content.Model;
+using GovUk.Education.ExploreEducationStatistics.Content.Model.Database;
+using GovUk.Education.ExploreEducationStatistics.Public.Data.Model;
+using GovUk.Education.ExploreEducationStatistics.Public.Data.Model.Database;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.EntityFrameworkCore;
+
+namespace GovUk.Education.ExploreEducationStatistics.Admin.Services.Public.Data;
+
+public class DataSetService(
+    ContentDbContext contentDbContext,
+    PublicDataDbContext publicDataDbContext,
+    IUserService userService)
+    : IDataSetService
+{
+    public async Task<Either<ActionResult, PaginatedListViewModel<DataSetViewModel>>> ListPublicationDataSets(
+        int page,
+        int pageSize,
+        Guid publicationId,
+        CancellationToken cancellationToken = default)
+    {
+        return await CheckPublicationExists(publicationId, cancellationToken)
+            .OnSuccess(userService.CheckCanViewPublication)
+            .OnSuccess(async () =>
+            {
+                var dataSetsQueryable = publicDataDbContext.DataSets
+                    .AsNoTracking()
+                    .Include(ds => ds.LatestLiveVersion)
+                    .Include(ds => ds.Versions)
+                    .Where(ds => ds.PublicationId == publicationId);
+
+                var dataSets = (await dataSetsQueryable
+                        .OrderByDescending(ds => ds.LatestLiveVersion != null ? ds.LatestLiveVersion.Published : DateTimeOffset.MinValue)
+                        .ThenBy(ds => ds.Title)
+                        .ThenBy(ds => ds.Id)
+                        .Paginate(page: page, pageSize: pageSize)
+                        .ToListAsync(cancellationToken: cancellationToken)
+                    )
+                    .Select(MapDataSet)
+                    .ToList();
+
+                return new PaginatedListViewModel<DataSetViewModel>(
+                    dataSets,
+                    totalResults: await dataSetsQueryable.CountAsync(cancellationToken: cancellationToken),
+                    page: page,
+                    pageSize: pageSize);
+            });
+    }
+
+    private static DataSetViewModel MapDataSet(DataSet dataSet)
+    {
+        var latestDraftVersion = dataSet.Versions
+            .Where(version => version.Status is DataSetVersionStatus.Draft
+                or DataSetVersionStatus.Failed
+                or DataSetVersionStatus.Mapping
+                or DataSetVersionStatus.Processing)
+            .OrderBy(version => version.VersionMajor)
+            .ThenBy(version => version.VersionMinor)
+            .LastOrDefault();
```

Comment:
It's a bit inefficient to have to go through every single version to figure out what the latest draft is.

It's also possible we might end up with lots of data sets and lots of versions for those data sets, so this approach isn't super scaleable. In fact, it's basically the same type of problem we had / have with releases.

Would suggest that rather than do this, we add `LatestDraftVersion` / `LatestDraftVersionId` to `DataSetVersion` so that we have pointers to these in the database.

Reply:
I've added this to `DataSet` now, squashed the migration and regenerated `public-api-7.zip`.